frontend/test/metabase/scenarios/permissions/download-permissions.cy.spec.js

```diff
@@ -88,51 +80,40 @@ describeEE("scenarios > admin > permissions", () => {
       cy.button("Yes").click();
     });
 
-    assertPermissionForItem(groupName, DOWNLOAD_PERMISSION_INDEX, "No");
+    assertPermissionForItem("All Users", DOWNLOAD_PERMISSION_INDEX, "No");
   });
 
   it("restricts users from downloading questions", () => {
-    cy.visit("/admin/permissions/data/database/1");
-
     // Restrict downloads for All Users
-    modifyPermission("All Users", DOWNLOAD_PERMISSION_INDEX, "No");
-
-    cy.button("Save changes").click();
-
-    modal().within(() => {
-      cy.findByText("Save permissions?");
-      cy.findByText("Are you sure you want to do this?");
-      cy.button("Yes").click();
+    cy.updatePermissionsGraph({
+      [ALL_USERS_GROUP]: {
+        [SAMPLE_DB_ID]: {
+          download: { schemas: "none" },
+        },
+      },
     });
 
     cy.signInAsNormalUser();
-    cy.visit("/question/1");
+    visitQuestion("1");
 
     cy.findByText("Showing first 2,000 rows");
     cy.icon("download").should("not.exist");
   });
 
   it("limits users from downloading all results", () => {
     const questionId = 1;
-    cy.visit("/admin/permissions/data/database/1");
 
     // Restrict downloads for All Users
-    modifyPermission("All Users", DOWNLOAD_PERMISSION_INDEX, "No");
-
-    // Limit downloads for "data" group
-    const groupName = "data";
-    modifyPermission(groupName, DOWNLOAD_PERMISSION_INDEX, "10 thousand rows");
-
-    cy.button("Save changes").click();
-
-    modal().within(() => {
-      cy.findByText("Save permissions?");
-      cy.findByText("Are you sure you want to do this?");
-      cy.button("Yes").click();
+    cy.updatePermissionsGraph({
```

Comment:
Doesn't that make the tests more isolating? What if someone introduces a bug which causes removal of "10 thousand rows" select item in the UI or a bug in a code that updates and sends the graph to the BE?

Reply:
You're right in this particular case because we didn't choose "10 thousand rows" using the UI previously. However, a lot of duplication can and should be avoided by using API (after we've tested something using the UI).

> What if someone introduces a bug which causes removal of "10 thousand rows" select item in the UI

Rather than going a full matrix style and checking every single option, we could easily have a visual regression test for that.

> or a bug in a code that updates and sends the graph to the BE?

That is a good question. We should expand our test coverage not only by piling more and more E2E tests, but also by introducing API tests and possibly contract testing.

The way we're currently going with E2E is unmaintainable in the long run. They are usually always at the very top of the "testing pyramid" and should be the smallest percentage in the total testing suite. We reversed that.